src/grpc_interceptor/client.py

```diff
@@ -0,0 +1,172 @@
+"""Base class for client-side interceptors."""
+
+import abc
+from collections import namedtuple
+from typing import Any, Callable, Iterator, Optional, Tuple
+
+import grpc
+
+
+class ClientCallDetails(
+    namedtuple(
+        "ClientCallDetails",
+        (
+            "method",
+            "timeout",
+            "metadata",
+            "credentials",
+            "wait_for_ready",
+            "compression",
+        ),
+    ),
+    grpc.ClientCallDetails,
+):
+    """Describes an RPC to be invoked.
+
+    This is an EXPERIMENTAL API.
+
+    Attributes:
+        method: The method name of the RPC.
+        timeout: An optional duration of time in seconds to allow for the RPC.
+        metadata: Optional :term:`metadata` to be transmitted to the
+                  service-side of the RPC.
+        credentials: An optional CallCredentials for the RPC.
+        wait_for_ready: This is an EXPERIMENTAL argument. An optional flag to
+                        enable :term:`wait_for_ready` mechanism.
+        compression: An element of grpc.compression, e.g. grpc.compression.Gzip.
+                     This is an EXPERIMENTAL option.
+    """
+
+    pass
+
+
+class ClientInterceptor(
+    grpc.UnaryUnaryClientInterceptor,
+    grpc.UnaryStreamClientInterceptor,
+    grpc.StreamUnaryClientInterceptor,
+    grpc.StreamStreamClientInterceptor,
+    metaclass=abc.ABCMeta,
+):
+    """Base class for client-side interceptors.
+
+    To implement an interceptor, subclass this class and override the intercept method.
+    """
+
+    @abc.abstractmethod
+    def intercept(
+        self,
+        call_details: ClientCallDetails,
+        request_iterator: Iterator[Any],
```

Comment:
It feels a little off to pass an iterator when the RPC is unary-unary. Since that's the most common type of RPC, I think  this could confuse folks. Any reason not to call this `request_or_iterator` and just pass through whatever `grpc` gives us? If the interceptor wants to modify the request and needs to know whether it should iterator over it, it can look at the `request_streaming` parameter.

Reply:
Reasonable, can do.

Reply:
Refactored to just pass in the continuation, details, and request message/iterator as-is per other suggestions. The parameter is named `request_or_iterator` and the interceptor would need to type check to determine which. I can add the bools back in if you think that's better but the end result is basically the same, additional logic being required to determine streaming vs. unary.